### ISSUE TITLE
Support "show in file manager" operation

### ISSUE THREAD

**Issue by probonopd** [helloSystem/Filer#15]
User story:
When I select "show in folder" for a download in a browser, then not only the Downloads folder should open but also the downloaded file should be selected.

Looks like GNOME and KDE use https://www.freedesktop.org/wiki/Specifications/file-manager-interface/, `org.freedesktop.FileManager1` DBus name and the `/org/freedesktop/FileManager1` object path. 

And e.g., Chromium implements it:
https://chromium.googlesource.com/chromium/src/+/master/chrome/browser/platform_util_linux.cc

Reference:
https://github.com/probonopd/go-appimage/issues/12

This was also requested upstream.

**Comment by probonopd** [comment]
https://github.com/linuxdeepin/dde-file-manager/search?q=dbusfilemanager1
seems to be relatively straightforward.

**Comment by probonopd** [comment]
![image](https://user-images.githubusercontent.com/2480569/92415090-d44c5b00-f146-11ea-9a68-7478c612a9f8.png)

**Comment by grahamperrin** [comment]
… also (for example) this **Open folder** button in [multimedia/simplescreenrecorder](https://www.freshports.org/multimedia/simplescreenrecorder/):

![image](https://user-images.githubusercontent.com/192271/106386683-13b56580-63ce-11eb-9e48-d851b006c894.png)**

**Comment by probonopd** [comment]
SSR is using https://doc.qt.io/qt-5/qdesktopservices.html#openUrl

https://github.com/MaartenBaert/ssr/blob/5f50ceff93cbdc1eab8ad09af880f5b8ffb28702/src/GUI/PageDone.cpp#L54-L57

> [static]bool QDesktopServices::openUrl(const QUrl &url)
> Opens the given url in the appropriate Web browser for the user's desktop environment, and returns true if successful; otherwise returns false.
> 
> If the URL is a reference to a local file (i.e., the URL scheme is "file") then it will be opened with a suitable application instead of a Web browser.

I think this is not about "show in file manager".

**Comment by grahamperrin** [comment]
With KDE Plasma on FreeBSD `14.0-CURRENT`, the **Open folder** button opens the folder that contains the screen recording in a new tab in Dolphin.

**Comment by probonopd** [comment]
> the Open folder button opens the folder that contains the screen recording in a new tab in Dolphin.

Yes, but this ticket is about applications being able to not only open the folder but also select the file in the file manager. Filer does not implement the protocol needed for this yet.

> Do you need a separate issue for applications that should show the folder but not select the file?

Not yet. This stuff is expected to work only once we have a proper "open with" database/implementation.

**Comment by probonopd** [comment]
Looks like @Ashish-Bansal had sent a patch for KDE Dolphin that implements this.

https://phabricator.kde.org/source/dolphin/browse/master/src/dbusinterface.cpp

Can we port this to Filer?

**Comment by probonopd** [comment]
Did I ever mention that I find D-Bus way too complicated? ;-)

Added
https://github.com/helloSystem/Filer/commit/4c1f261220f9d0993b219c32f57a929aec753b80

but the wrong methods get exported to D-Bus under `org.freedesktop.FileManager1`:
* Why is there `Application/org.filer.application` with methods like `desktopPreferences`?
* Why is there not `/org/freedesktop/FileManager1` with methods like `ShowFolders`?

**Comment by probonopd** [comment]
What I seem to be missing entirely is how to tell Qt "publish `ShowFolders` under `/org/freedesktop/FileManager1` under `org.freedesktop.FileManager1`. Exactly there and only there.

**Comment by probonopd** [comment]
Thanks @zyga for the following explanation:

> I don't know Qt but it's fairly simple if you think about it the following way.

- Any client can connect to the bus and publish objects with paths.
- Any object can offer one or more interface.
- All methods and properties belong to an interface.
- A client can claim a bus name.
- Anyone trying to find a given bus name, object path and interface can trigger bus activation, starting the providing service.
- Interface is a personality. Each object can present arbitrary many. That’s the idea. Method Foo and property Bar are only meaningful if you add the interface (personality) name to it. You can have any number of methods Foo as long as the interface is different.

Maybe if I meditate long enough over it, then one day I will undestand it ;-)

**Comment by probonopd** [comment]
This works:

`gdbus call --session --dest org.freedesktop.FileManager1 --object-path /org/freedesktop/FileManager1  --method org.freedesktop.FileManager1.ShowItems '["file:///Applications"]' ""`

It will open a Filer window at / and highlight the Applications folder.

Thanks @mszoek.

**Comment by probonopd** [comment]
Looks like `ShowFolders` is broken.

Causes:
* https://github.com/helloSystem/ISO/issues/554#issuecomment-1807183444
